fabric-data-generation-api-v1/src/client/java/net/fabricmc/fabric/api/client/datagen/v1/builder/SoundTypeBuilder.java

```diff
@@ -0,0 +1,213 @@
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.client.datagen.v1.builder;
+
+import java.util.Objects;
+
+import com.mojang.serialization.Codec;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.sound.SoundCategory;
+import net.minecraft.sound.SoundEvent;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.StringIdentifiable;
+import net.minecraft.util.Util;
+
+import net.fabricmc.fabric.api.client.datagen.v1.provider.FabricSoundsProvider;
+import net.fabricmc.fabric.impl.datagen.client.SoundTypeBuilderImpl;
+
+/**
+ * Provides functionality for building entries that go into a {@code sounds.json} file.
+ *
+ * <p>Use in conjunction with {@link FabricSoundsProvider} to generate sound definitions.
+ *
+ * @see net.minecraft.client.sound.SoundManager
+ * @see net.minecraft.client.sound.WeightedSoundSet
+ */
+@ApiStatus.NonExtendable
+public interface SoundTypeBuilder {
+	/**
+	 * Creates a new builder pre-filled with a subtitle translation string based on the passed event.
+	 *
+	 * @return New sound type builder
+	 */
+	static SoundTypeBuilder of(SoundEvent event) {
+		Objects.requireNonNull(event, "Sound event cannot be null.");
+		return of().subtitle(Util.createTranslationKey("subtitles", event.id()));
+	}
+
+	/**
+	 * Creates a new empty builder.
+	 *
+	 * @return New sound type builder
+	 */
+	static SoundTypeBuilder of() {
+		return new SoundTypeBuilderImpl();
+	}
+
+	/**
+	 * Sets the sound category the sound event must play on.
+	 *
+	 * <p>The default category is {@link SoundCategory#NEUTRAL}. GUI elements should use {@link SoundCategory#MASTER}.
+	 */
+	SoundTypeBuilder category(SoundCategory category);
+
+	/**
+	 * Sets an optional translation string to use for the sound's subtitle.
+	 *
+	 * <p>The default is null (no subtitle).
+	 */
+	SoundTypeBuilder subtitle(@Nullable String subtitle);
+
+	/**
+	 * Adds one sound to the event.
+	 *
+	 * @param sound base sound to add
+	 */
+	SoundTypeBuilder sound(EntryBuilder sound);
+
+	/**
+	 * Adds one or more sounds to the event.
+	 *
+	 * <p>This is a shorthand method for quickly adding multiple
+	 * entries where each sound is a variant with an index at the end of their name.
+	 *
+	 * <p>Calling this with the a count value of {@code 3} is the equivalent of doing:
+	 *
+	 * <p>{@code builder.sound(EntryBuilder.ofFile(id.withSuffixedPath("1"));}
+	 * {@code builder.sound(EntryBuilder.ofFile(id.withSuffixedPath("2"));}
+	 * {@code builder.sound(EntryBuilder.ofFile(id.withSuffixedPath("3"));}
+	 *
+	 * @param sound base sound to add
+	 * @param count number of instances of that sound to register
+	 */
+	SoundTypeBuilder sound(EntryBuilder sound, int count);
+
+	/**
+	 * Represents the type of weighted sound entry.
+	 *
+	 * @see net.minecraft.client.sound.Sound.RegistrationType
+	 */
+	enum RegistrationType implements StringIdentifiable {
```

Comment:
`Codec.STRING.xmap(RegistrationType::getByName, t -> t.name)` is surely sufficient, no?

Reply:
RegistrationType doesn't have a getByName method. RegistrationType#name is private.

Like I said in a previous comment, I can change it to use the regular RegistrationType but I'd like for the comments from this one to be added to yarn.

Reply:
I think this is fine to duplicate, as the entry and sound type are. There is very little benefit in using the vanilla enum.